### PR TITLE
Add NAT GW for private subnets

### PR DESCRIPTION
**Pull request by prateekgogia**
Issue #, if available:
https://github.com/awslabs/kubernetes-iteration-toolkit/issues/195

Description of changes:
Adds logic to create NAT GW in KIT environment
```
Created NAT gateway fargate-testing-pgogia, ID nat-008e25c7ab60fb0f2
Ensured launch template version 3 for fargate-testing-pgogia
Found instance i-0318a7af943f4239a
NAT gateway is available fargate-testing-pgogia
To access cluster run -

 	export KUBECONFIG=/tmp/fargate-testing-pgogia/etc/kubernetes/admin.conf

Installed chart https://aws.github.io/eks-charts/aws-vpc-cni
Applied https://storage.googleapis.com/tekton-releases/pipeline/previous/v0.33.2/release.yaml
Applied https://storage.googleapis.com/tekton-releases/triggers/previous/v0.19.0/release.yaml
Applied https://github.com/tektoncd/dashboard/releases/download/v0.24.1/tekton-dashboard-release.yaml
Installed chart https://aws.github.io/eks-charts/aws-load-balancer-controller
Installed chart https://awslabs.github.io/kubernetes-iteration-toolkit/kit-operator
Installed chart https://charts.karpenter.sh/karpenter
Installed chart https://github.com/kubernetes-sigs/aws-ebs-csi-driver/releases/download/helm-chart-aws-ebs-csi-driver-2.6.3/aws-ebs-csi-driver
Installed chart https://github.com/prometheus-community/helm-charts/releases/download/kube-prometheus-stack-34.0.0//kube-prometheus-stack
✅ Bootstrapped "fargate-testing-pgogia" after 4m34.347270336s
```

Tested delete for the environment-

```
Deleted NAT gateway fargate-testing-pgogia
Deleted instances [i-0318a7af943f4239a]
Deleted role fargate-testing-pgogia
Deleted policy fargate-testing-pgogia-tenant-controlplane-node-role from role fargate-testing-pgogia-tenant-controlplane-node-role
Deleted association of route table rtb-0d4930d26a39b8872 to subnet subnet-04c6c3f905afc3032
Deleted policy fargate-testing-pgogia-tenant-controlplane-node-role from role fargate-testing-pgogia-tenant-controlplane-node-role
Deleted policy fargate-testing-pgogia-tenant-controlplane-node-role from role fargate-testing-pgogia-tenant-controlplane-node-role
Deleted policy fargate-testing-pgogia-tenant-controlplane-node-role from role fargate-testing-pgogia-tenant-controlplane-node-role
Deleted association of route table rtb-0d4930d26a39b8872 to subnet subnet-0455dea2394e4791e
Deleted policy fargate-testing-pgogia-tenant-controlplane-node-role from role fargate-testing-pgogia-tenant-controlplane-node-role
Deleted policy fargate-testing-pgogia-tenant-controlplane-node-role from role fargate-testing-pgogia-tenant-controlplane-node-role
Deleted role fargate-testing-pgogia-tenant-controlplane-node-role from instance profile fargate-testing-pgogia-tenant-controlplane-node-role
Deleted instance profile fargate-testing-pgogia-tenant-controlplane-node-role
Deleted role fargate-testing-pgogia-tenant-controlplane-node-role
Deleted association of route table rtb-0d4930d26a39b8872 to subnet subnet-08851360a83b867bd
Deleted route table rtb-0db488634da5616d7
Deleted route table rtb-0d4930d26a39b8872
Deleted S3 bucket fargate-testing-pgogia
Deleted security group sg-00f54e5d371f44f4f
Deleted internet gateway vpc-0c40f9715c17df35c attachment to igw-0a0c6e9a6d11d9e68
Deleted subnet subnet-0455dea2394e4791e
Deleted internet gateway igw-0a0c6e9a6d11d9e68
Deleted subnet subnet-0455dea2394e4791e
Deleted subnet subnet-0455dea2394e4791e
Deleted subnet subnet-0455dea2394e4791e
Deleted vpc vpc-0c40f9715c17df35c
Deleted substrate fargate-testing-pgogia after 49.60391238s
```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: substrate/pkg/controller/substrate/infrastructure/natgateway.go [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+		len(substrate.Status.Infrastructure.PublicSubnetIDs) == 0 ||
+		substrate.Status.Infrastructure.ElasticIpIDForNatGW == nil {
+		return reconcile.Result{Requeue: true}, nil
+	}
+	natGW, err := n.ensure(ctx, substrate)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	if _, err := n.EC2.CreateRouteWithContext(ctx, &ec2.CreateRouteInput{
+		RouteTableId:         substrate.Status.Infrastructure.PrivateRouteTableID,
```
Comment (hakuna-matatah) [inline substrate/pkg/controller/substrate/infrastructure/natgateway.go:45]: does this route table contains all the private subnets cidrs ? Just want to make sure every private subnet has a route to NATGW.

File: substrate/pkg/controller/substrate/infrastructure/natgateway.go [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+	if len(output.NatGateways) == 0 {
+		return nil, nil
+	}
+	var result *ec2.NatGateway
+	for _, natgw := range output.NatGateways {
+		if aws.StringValue(natgw.State) == "deleting" || aws.StringValue(natgw.State) == "deleted" ||
+			aws.StringValue(natgw.State) == "failed" {
+			continue
+		}
+		if result != nil {
```
Comment (hakuna-matatah) [inline substrate/pkg/controller/substrate/infrastructure/natgateway.go:70]: Looks like we fail if we find more than one NATs in VPC ,  we can have a `Todo` for future (nice to have) to add atleast 2 NATs so we are immune to AZ failures for all clusters using KIT Environment :D

File: substrate/pkg/controller/substrate/cluster/address.go [hunk 3]
```diff
@@ -30,28 +33,52 @@ type Address struct {
...
+	ipAddress    *string
+	allocationID *string
+}
+
 func (a *Address) Create(ctx context.Context, substrate *v1alpha1.Substrate) (reconcile.Result, error) {
-	addressesOutput, err := a.EC2.DescribeAddressesWithContext(ctx, &ec2.DescribeAddressesInput{Filters: discovery.Filters(substrate, discovery.Name(substrate))})
+	for _, req := range []struct {
+		name *string
+	}{
+		{name: discovery.Name(substrate, "apiserver")},
```
Comment (hakuna-matatah) [inline substrate/pkg/controller/substrate/cluster/address.go:45]: we might want to document this behavior ? so customers ensure they don't duplicate these tags like `apiserver` and `natgw` outside of KIT Environment as you're discovering them below with these filters.

File: substrate/pkg/controller/substrate/infrastructure/natgateway.go (hunk 1 above)
Comment (prateekgogia) [reply substrate/pkg/controller/substrate/infrastructure/natgateway.go:46]: Yes all the subnets share the same NAT GW and all of them are added to the route table.

File: substrate/pkg/controller/substrate/infrastructure/natgateway.go (hunk 2 above)
Comment (prateekgogia) [reply substrate/pkg/controller/substrate/infrastructure/natgateway.go:73]: yeah sure once we add HA to the control plane.

**Pull request merged**